dbms/src/Flash/Coprocessor/DAGRequest.cpp

```diff
@@ -0,0 +1,137 @@
+        case tipb::ExecType::TypeStreamAgg:
+        case tipb::ExecType::TypeAggregation:
+            return fmt::format("aggregation_{}", current_id++);
+        case tipb::ExecType::TypeTopN:
+            return fmt::format("topn_{}", current_id++);
+        case tipb::ExecType::TypeLimit:
+            return fmt::format("limit_{}", current_id++);
+        case tipb::ExecType::TypeExchangeSender:
+            return fmt::format("exchange_sender_{}", current_id++);
+        case tipb::ExecType::TypeExchangeReceiver:
+            return fmt::format("exchange_receiver_{}", current_id++);
+        case tipb::ExecType::TypeTableScan:
+        case tipb::ExecType::TypePartitionTableScan:
+            return fmt::format("table_scan_{}", current_id++);
+        case tipb::ExecType::TypeSort:
+            return fmt::format("sort_{}", current_id++);
+        case tipb::ExecType::TypeWindow:
+            return fmt::format("window_{}", current_id++);
+        case tipb::ExecType::TypeJoin:
+            return fmt::format("join_{}", current_id++);
+        case tipb::ExecType::TypeExpand:
+            return fmt::format("expand_{}", current_id++);
+        default:
+            throw TiFlashException(
+                fmt::format("Unsupported executor in DAG request: {}", executor.DebugString()),
+                Errors::Coprocessor::Unimplemented);
+        }
+    }
+
+private:
+    UInt32 current_id = 0;
+};
+} // namespace
+
+DAGRequest::DAGRequest(tipb::DAGRequest * dag_request_)
+    : dag_request(dag_request_)
+{
+    // Will only occur in tests.
+    if unlikely (!dag_request)
+        return;
+
+    check((dag_request->executors_size() > 0) != dag_request->has_root_executor(), "dagrequest must be one of list based and tree based");
+    is_tree_based = dag_request->has_root_executor();
+
+    checkOrSetExecutorId();
+}
+
+void DAGRequest::checkOrSetExecutorId()
+{
+    if (is_tree_based)
+    {
+        // check duplicate executor_id for tree based request.
+        std::unordered_set<String> ids;
+        traverseExecutorTree(dag_request->root_executor(), [&](const tipb::Executor & executor) {
+            check(executor.has_executor_id(), "for tree based request, executor id cannot be null");
+            auto executor_id = executor.executor_id();
+            check(ids.find(executor_id) == ids.end(), fmt::format("in tree based request, executor id `{}` duplicate, which is unexpected.", executor_id));
+            ids.insert(executor_id);
+            return true;
+        });
+    }
+    else
+    {
+        // generate executor_id for list based request.
+        ExecutorIdGenerator id_generator;
+        for (int i = 0; i < dag_request->executors_size(); ++i)
+        {
+            auto * executor = dag_request->mutable_executors(i);
+            check(!executor->has_executor_id(), fmt::format("for list based request, executor id must be null, the unexpected executor id: {}", executor->executor_id()));
```

Comment:
Not sure if it is safe.
```
// for disaggregated task on write node
DAGContext::DAGContext(tipb::DAGRequest & dag_request_, const disaggregated::DisaggTaskMeta & task_meta_, TablesRegionsInfo && tables_regions_info_, const String & compute_node_host_, LoggerPtr log_)
    : dag_request(&dag_request_)
    , dummy_query_string(dag_request->DebugString())
    , dummy_ast(makeDummyQuery())
    , tidb_host(compute_node_host_)
    , collect_execution_summaries(dag_request->has_collect_execution_summaries() && dag_request->collect_execution_summaries())
    , is_mpp_task(false)
    , is_root_mpp_task(false)
    , is_batch_cop(false)
    , is_disaggregated_task(true)
    , tables_regions_info(std::move(tables_regions_info_))
    , log(std::move(log_))
    , flags(dag_request->flags())
    , sql_mode(dag_request->sql_mode())
    , disaggregated_id(std::make_unique<DM::DisaggTaskId>(task_meta_))
    , max_recorded_error_count(getMaxErrorCount(*dag_request))
    , warnings(max_recorded_error_count)
    , warning_count(0)
    , keyspace_id(RequestUtils::deriveKeyspaceID(task_meta_))
{
    RUNTIME_CHECK(isTreeBasedExecutors() && dag_request->root_executor().has_executor_id());
    return_executor_id = dag_request->root_executor().has_executor_id() || dag_request->executors(0).has_executor_id();

    initOutputInfo();
    initListBasedExecutors();
}
```
It assumes that list-based executor may have `executor_id`

Reply:
From https://github.com/pingcap/tidb/blob/3c942f7e101d6b0aff85c552598678f2a761deca/planner/core/plan_to_pb.go#L37 and https://github.com/pingcap/tispark/blob/c908c5516ea70c17b1bab5c0de3bafacb5b5c2a1/tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java#L234, executor_id will be empty in list base request.
But I think it's fine to not require the list based executor_id to be empty.
/cc @windtalker

Reply:
But IMO the request is send from TiFlash CN to DN, and we have changed the DAGRequest already.(I am not sure)

Reply:
https://github.com/pingcap/tidb/blob/3c942f7e101d6b0aff85c552598678f2a761deca/planner/core/plan_to_pb.go#L160-L169
It seems to be a bug that constructs a list based request with an executor id.
Although this request will only be sent to tikv...

Reply:
https://github.com/pingcap/tiflash/pull/7124/files#diff-759cb48a99e0dc5922fc34aa80a79f1ef4b99e6986deba493d993e59cbba99ffR29-R67
No longer assume that a list based executor must not have an executor id.